README.md
# ESPHome CANopen external component

This project provides ESPHome external_component named `canopen` for easy conversion of ESPHome entities into CANopen nodes. It allows to map selected entities to CANopen object dictionary. It also allows to define TPDOs for sending entity state on change. Currently following entity types are supported: `sensor`, `binary_sensor`, `switch`, `cover` and `light`. Entity metadata is also published in Object Dictionary enabling autodiscovery.

The [can2mqtt](https://github.com/mrk-its/can2mqtt) project provides can2mqtt bridge exposing ESPHome CANopen entities onto MQTT topics. It follows MQTT Discovery protocol, so entities appear automatically in HomeAssistant.

# How to start
## Hardware requirements
 * any ESP8266/ESP32 board with CAN controller. ESP32 boards are preferred, as they have [integrated can controller](https://esphome.io/components/canbus.html#esp32-can-component) (but still cheap external CAN transceiver is needed). For ESP8266 external [MCP2515](https://esphome.io/components/canbus.html#mcp2515-component) CAN controller can be used.
## Software requirements
 * [ESPHome](https://esphome.io/) installed

## Configure ESPHome with `canopen` external_component

  Example configuration for esp32dev board:

  ```
  esphome:
    name: can-node-1

  external_components:
    - source: github://mrk-its/esphome-canopen

  # Enable logging
  logger:

  # Enable Home Assistant API
  api:
    password: ""
    reboot_timeout: 0s

  ota:
    password: ""

  esp32:
    board: esp32dev
    framework:
      type: arduino

  wifi:
    <<: !include {file: wifi.yaml}

  canbus:
    - platform: esp32_can
      id: can_bus
      rx_pin: GPIO22
      tx_pin: GPIO23
      can_id: 0
      bit_rate: 125kbps

  canopen:
    id: can_gate
    canbus_id: can_bus
    node_id: 1
    entities:
      - id: boot
        index: 1
        tpdo: 0
      - id: blue_led
        index: 2
        tpdo: 0
      - id: uptime_sensor
        index: 3
        tpdo: 0
      - id: cover1
        index: 4
        tpdo: 1
      - id: cover2
        index: 5
        tpdo: 1

  sensor:
    - platform: uptime
      id: uptime_sensor
      name: "Uptime 1"
      update_interval: 5sec
      internal: true

  binary_sensor:
    - platform: gpio
      name: "Boot 1"
      id: boot
      internal: true
      pin:
        number: 0
        inverted: true

  switch:
    - platform: gpio
      name: "Led 1"
      id: blue_led
      internal: true
      pin: 2

  cover:
    - platform: time_based
      name: "Cover 1"
      id: cover1
      internal: true
      device_class: shutter
      has_built_in_endstop: true
      open_action:
        - logger.log: open_action
      open_duration: 10s
      close_action:
        - logger.log: close_action
      close_duration: 10s
      stop_action:
        - logger.log: stop_action

  ```

# Support
## Community

* [Discord Server](https://discord.gg/VXjUSnUWsd)


CHANGELOG.md
# 2024-05-27, v0.3.0
* add support for heartbeat consumers
* fix serious timer driver bug


# 2023-05-23, v0.2.1
* add support for up to 8 TPDOs (TPDO 4..7 use RPDO 0..3 CAN ids)
* heartbeat interval is configurable via canopen.heartbeat_interval configuration property
* `state` enum added to `status_info` struct
* new methods: `initiate_recovery()` and `start()` for performing recovery from bus_error state
* esphome.name exposed as manufacturer device name (OD: 0x1008)
* hw_version (OD: 0x1009) and sw_version (OD: 0x100a) configurable via `hw_version` / `sw_version` config properties
* all esphome-devices have the same Ientity VendorId / ProductCode magic values (used by can2mqtt)
* added support for Store Parameters (OD: 0x1010) and Restore default parameters (OD: 0x1011), used for storing RPDO configuration

# 2023-05-13, v0.1.0
* support for packing sensor data into 8-bit or 16-bit unsigned integers

  Is it possible now to scale and pack sensor value into 8-bit / 16-bit unsigned integer, to reduce size in TPDO frame. Sensor value is scaled to 8-bit or 16-bit width before transmission and scaled back on receive. Following additional parameters are recognized:
  * size (in bytes), valid values are 1 (8-bit int) or 2 (16-bit int)
  * min_value: min_value of sensor, used for scaling
  * max_value: max value of sensor, used for scaling
  values outside of range are clipped into [min_value, max_value]

* configure canopen early, with setup_priority::HARDWARE
* fix 'oscillations' caused by mutual state updates by delaying state update by configured amount of time (50ms by default, can be changed via `state_update_delay` config option)
* add `trig_tpdo` method to trigger TPDO


components/canopen/__init__.py
from itertools import groupby
import esphome.config_validation as cv
import esphome.codegen as cg
from esphome import automation
from esphome.const import CONF_ID, CONF_TRIGGER_ID
from esphome.components.canbus import CanbusComponent


ns = cg.esphome_ns.namespace('canopen')
CanopenComponent = ns.class_(
    'CanopenComponent',
    cg.Component,
)

PreOperationalTrigger = ns.class_("PreOperationalTrigger", automation.Trigger.template())
OperationalTrigger = ns.class_("OperationalTrigger", automation.Trigger.template())
HbConsumerEventTrigger = ns.class_("HbConsumerEventTrigger", automation.Trigger.template())

CONF_ENTITIES = "entities"

DEPENDENCIES = ["ota"]

CSDO_SCHEMA = cv.Schema({
    cv.Required("node_id"): cv.int_,
    cv.Optional("tx_id", 0x600): cv.int_,
    cv.Optional("rx_id", 0x580): cv.int_,
})

RPDO_SCHEMA = cv.Schema({
    cv.Required("node_id"): cv.int_,
    cv.Required("tpdo"): cv.int_,
    cv.Required("offset"): cv.int_,
    cv.Optional("cmd", 0): cv.int_
})

ENTITY_SCHEMA = cv.Schema({
    cv.Required("id"): cv.use_id(cg.EntityBase),
    cv.Required("index"): cv.int_,
    cv.Optional("size"): cv.int_,
    cv.Optional("min_value", 0): cv.float_,
    cv.Optional("max_value", 0): cv.float_,
    cv.Optional("tpdo"): cv.int_,
    cv.Optional("rpdo"): cv.ensure_list(RPDO_SCHEMA),
})
HB_CLIENT_SCHEMA = cv.Schema({
    cv.Required("node_id"): cv.int_,
    cv.Required("timeout"): cv.positive_time_period_milliseconds,
})

CONFIG_SCHEMA = cv.Schema({
    cv.GenerateID(): cv.declare_id(CanopenComponent),
    cv.Required("canbus_id"): cv.use_id(CanbusComponent),
    cv.Required("node_id"): cv.int_,
    # cv.Optional("status"): STATUS_ENTITY_SCHEMA,
    cv.Optional("csdo"): cv.ensure_list(CSDO_SCHEMA),
    cv.Required(CONF_ENTITIES): cv.ensure_list(ENTITY_SCHEMA),
    cv.Optional("on_pre_operational"): automation.validate_automation({
        cv.GenerateID(CONF_TRIGGER_ID): cv.declare_id(PreOperationalTrigger),
    }),
    cv.Optional("on_operational"): automation.validate_automation({
        cv.GenerateID(CONF_TRIGGER_ID): cv.declare_id(OperationalTrigger),
    }),
    cv.Optional("on_hb_consumer_event"): automation.validate_automation({
        cv.GenerateID(CONF_TRIGGER_ID): cv.declare_id(HbConsumerEventTrigger),
    }),
    cv.Optional("state_update_delay", "50ms"): cv.positive_time_period_microseconds,
    cv.Optional("heartbeat_interval", "5000ms"): cv.positive_time_period_milliseconds,
    cv.Optional("heartbeat_clients"): cv.ensure_list(HB_CLIENT_SCHEMA),
    cv.Optional("sw_version"): cv.string,
    cv.Optional("hw_version"): cv.string
}).extend(cv.COMPONENT_SCHEMA)

def to_code(config):
    cg.add_library("canopenstack=https://github.com/mrk-its/canopen-stack#dev", "0.0.0")
    # cg.add_library("canopenstack=file:///home/mrk/canopen-stack", "0.0.0")
    canbus = yield cg.get_variable(config["canbus_id"])
    node_id = config["node_id"]
    canopen = cg.new_Pvariable(config[CONF_ID], canbus, node_id)
    cg.add(canopen.set_state_update_delay(config["state_update_delay"]))
    cg.add(canopen.set_heartbeat_interval(config["heartbeat_interval"]))
    hw_version = config.get("hw_version")
    sw_version = config.get("sw_version")
    if hw_version:
        cg.add(canopen.od_set_string(0x1009, 0, hw_version))
    if sw_version:
        cg.add(canopen.od_set_string(0x100a, 0, sw_version))

    yield cg.register_component(canopen, config)

    entities = sorted(config.get(CONF_ENTITIES, []), key=lambda x: x['index'])
    assert len(entities) == len(set(e['index'] for e in entities)), "All entity indices must be unique!"
    for entity_config in entities:
        entity = yield cg.get_variable(entity_config["id"])
        tpdo = entity_config.get("tpdo", -1)
        size = entity_config.get("size")
        if size in (1, 2):
            min_val = entity_config.get("min_value", 0)
            max_val = entity_config.get("max_value", 255 if size == 1 else 65535)
            cg.add(canopen.add_entity(entity, entity_config["index"], tpdo, size, min_val, max_val))
        else:
            cg.add(canopen.add_entity(entity, entity_config["index"], tpdo))

    for num, csdo in enumerate(config.get("csdo", ())):
        cg.add(canopen.setup_csdo(num, csdo["node_id"], csdo["tx_id"], csdo["rx_id"]))
    for num, client in enumerate(config.get("heartbeat_clients", ()), 1):
        cg.add(canopen.setup_heartbeat_client(num, client["node_id"], client["timeout"]))

    for conf in config.get("on_operational", []):
        trigger = cg.new_Pvariable(conf[CONF_TRIGGER_ID])
        yield automation.build_automation(trigger, [], conf)
        cg.add(canopen.add_trigger(trigger))

    for conf in config.get("on_pre_operational", []):
        trigger = cg.new_Pvariable(conf[CONF_TRIGGER_ID])
        yield automation.build_automation(trigger, [], conf)
        cg.add(canopen.add_trigger(trigger))

    for conf in config.get("on_hb_consumer_event", []):
        trigger = cg.new_Pvariable(conf[CONF_TRIGGER_ID])
        yield automation.build_automation(trigger, [(cg.uint8, "node_id"),], conf)
        cg.add(canopen.add_trigger(trigger))


    rpdo_entities = [
        {**rpdo, "entity_index": entity["index"]}
        for entity in entities
        for rpdo in entity.get("rpdo", ())
    ]
    rpdo_entities.sort(key=lambda rpdo: (rpdo["node_id"], rpdo["tpdo"], rpdo["offset"]))
    for idx, ((node_id, tpdo), rpdos) in enumerate(
        groupby(rpdo_entities, key=lambda rpdo: (rpdo["node_id"], rpdo["tpdo"]))
    ):
        cg.add(canopen.add_rpdo_node(idx, node_id, tpdo))
        rpdos = list(rpdos)
        curr_offs = 0
        for rpdo in rpdos:
            assert rpdo["offset"] >= curr_offs, f"RPDO: invalid TPDO offset {rpdo}"
            if rpdo["offset"] > curr_offs:
                cg.add(canopen.add_rpdo_dummy(idx, rpdo["offset"] - curr_offs))
                curr_offs = rpdo["offset"]
            cg.add(canopen.add_rpdo_entity_cmd(idx, rpdo["entity_index"], rpdo["cmd"]))
            curr_offs += 1
